math/fastor/include/algebra/math/impl/fastor_transform3.hpp

```diff
@@ -0,0 +1,234 @@
+
+  /** Constructor with arguments: translation
+   *
+   * @param t is the translation
+   **/
+  ALGEBRA_HOST
+  transform3(const vector3 &t) {
+
+    _data(Fastor::fseq<0, 3>(), 3) = t;
+
+    _data_inv = Fastor::inverse(_data);
+  }
+
+  /** Constructor with arguments: matrix
+   *
+   * @param m is the full 4x4 matrix
+   **/
+  ALGEBRA_HOST
+  transform3(const matrix44 &m) {
+    _data = m;
+
+    _data_inv = Fastor::inverse(_data);
+  }
+
+  /** Constructor with arguments: matrix as std::aray of scalar
+   *
+   * @param ma is the full 4x4 matrix asa 16 array
+   **/
+  ALGEBRA_HOST
+  transform3(const array_type<16> &ma) {
+    _data = m;
+
+    _data_inv = Fastor::inverse(_data);
+  }
+
+  /** Default contructors */
+  transform3() = default;
+  transform3(const transform3 &rhs) = default;
+  ~transform3() = default;
+
+  /** Equality operator */
+  ALGEBRA_HOST
+  inline bool operator==(const transform3 &rhs) const {
+
+    return _data == rhs._data;
+  }
+
+  /** This method retrieves the rotation of a transform */
+  ALGEBRA_HOST
+  inline auto rotation() const {
+
+    return Fastor::Tensor<scalar_t, 3, 3>(
+        _data(Fastor::fseq<0, 4>(), Fastor::fseq<0, 4>()));
+  }
+
+  /** This method retrieves x axis */
+  ALGEBRA_HOST_DEVICE
+  inline point3 x() const { return _data(Fastor::fseq<0, 3>(), 0); }
+
+  /** This method retrieves y axis */
+  ALGEBRA_HOST_DEVICE
+  inline point3 y() const { return _data(Fastor::fseq<0, 3>(), 1); }
+
+  /** This method retrieves z axis */
+  ALGEBRA_HOST_DEVICE
+  inline point3 z() const { return _data(Fastor::fseq<0, 3>(), 2); }
+
+  /** This method retrieves the translation of a transform */
+  ALGEBRA_HOST
+  inline vector3 translation() const { return _data(Fastor::fseq<0, 3>(), 3); }
+
+  /** This method retrieves the 4x4 matrix of a transform */
+  ALGEBRA_HOST
+  inline matrix44 matrix() const { return _data; }
+
+  /** This method retrieves the 4x4 matrix of an inverse transform */
+  ALGEBRA_HOST
+  inline matrix44 matrix_inverse() const { return _data_inv; }
+
+  /// start editing from here
+  /** This method transform from a point from the local 3D cartesian frame to
+   * the global 3D cartesian frame */
+  ALGEBRA_HOST
+  inline const point3 point_to_global(const point3 &v) const {
+
+    Fastor::Tensor<scalar_type, 4> vector_4;
+    vector_4(Fastor::fseq<0, 3>()) = v;
+    vector_4[3] = static_cast<scalar_type>(1);
+    return Fastor::Tensor<scalar_type, 3>(
+        Fastor::matmul(_data, vector_4)(Fastor::fseq<0, 3>()));
+  }
+
+  /** This method transform from a vector from the global 3D cartesian frame
+   * into the local 3D cartesian frame */
+  ALGEBRA_HOST
+  inline const point3 point_to_local(const point3 &v) const {
+
+    Fastor::Tensor<scalar_type, 4> vector_4;
+    vector_4(Fastor::fseq<0, 3>()) = v;
+    vector_4[3] = static_cast<scalar_type>(1);
+    return Fastor::Tensor<scalar_type, 3>(
+        Fastor::matmul(_data, vector_4)(Fastor::fseq<0, 3>()));
```

Comment:
I think, this should use the inverse transform ```_data_inv```.

Reply:
Fixed.